Add a DHT Concept to docs.ipfs.io (#240)
ipfs/docs: content/guides/concepts/dht.md

---
title: "Distributed Hash Table (DHT)"
menu:
    guides:
        parent: concepts
---

> # DHTs

Author: WIP specs are here https://github.com/libp2p/specs/blob/ee9734bdfa963709ab7346d5015d59d8d9ed65f4/kad-dht/README.md

---

> title: "Distributed Hash Table (DHT)"

Reviewer: ```suggestion
title: "Distributed Hash Tables (DHT)"
```

Reviewer: Moved this change to https://github.com/ipfs/docs/pull/364

menu:
    guides:
        parent: concepts
---

# DHTs

## What is a DHT?

> [Distributed Hash Tables](https://en.wikipedia.org/wiki/Distributed_hash_table) (DHT) are a distributed key-value store where keys are [cryptographic hashes](https://docs.ipfs.io/guides/concepts/hashes/). 

Reviewer: ```suggestion
[Distributed Hash Tables](https://en.wikipedia.org/wiki/Distributed_hash_table) (DHTs) are a distributed key-value store where keys are [cryptographic hashes](https://docs.ipfs.io/guides/concepts/hashes/). 
```

Reviewer: Moved this change to https://github.com/ipfs/docs/pull/364

DHTs are distributed. Each "peer" (or "node") is responsible for a subset of the DHT. 

> A receiving peer either answers the request, or the request is passed to another peer until a peer can answer it.

Reviewer: ```suggestion
When a peer receives a request, it either answers it, or the request is passed to another peer until a peer that can answer it is found.
```

Reviewer: Moved this change to https://github.com/ipfs/docs/pull/364

> Depending on implementations, a request not answered by the first contacted node can be :

Reviewer: ```suggestion
Depending on the implementation, a request not answered by the first node contacted can be:
```

Reviewer: Moved this change to https://github.com/ipfs/docs/pull/364

> - forwarded from peer to peer and the last peer contact the requesting peer

Reviewer: ```suggestion
- forwarded from peer to peer, with the last peer contacting the requesting peer
```

Reviewer: Moved this change to https://github.com/ipfs/docs/pull/364

> - forwarded from peer to peer and the answered is forwarded following the same path

Reviewer: ```suggestion
- forwarded from peer to peer, with the answer forwarded following the same path
```

Reviewer: Moved this change to https://github.com/ipfs/docs/pull/364

> - answered with contacting information of a node having better chances to be able to answer (IPFS uses this strategy)

Reviewer: ```suggestion
- answered with the contact information of a node that has better chances to be able to answer. **IPFS uses this strategy.**
```

Reviewer: Moved this change to https://github.com/ipfs/docs/pull/364

> DHT's decentralization provides advantages compared to a classic Key-Value store:

Reviewer: ```suggestion
DHTs' decentralization provides advantages compared to a traditional key-value store, including:
```

Reviewer: Moved this change to https://github.com/ipfs/docs/pull/364

> - *scalability* as a request for a hash of length *n* takes at most *log2(n)* steps to resolve.

Reviewer: ```suggestion
- *scalability*, since a request for a hash of length *n* takes at most *log2(n)* steps to resolve
```

Reviewer: Moved this change to https://github.com/ipfs/docs/pull/364

> - *fault tolerance* via redundancy, so that lookups are possible even if peers unexpectedly leave or join the DHT. Additionally, requests can be addressed to any peer if one is slow or unavailable.

Reviewer: ```suggestion
- *fault tolerance* via redundancy, so that lookups are possible even if peers unexpectedly leave or join the DHT. Additionally, requests can be addressed to any peer if another peer is slow or unavailable.
```

Reviewer: Moved this change to https://github.com/ipfs/docs/pull/364

> - *load balancing* as requests are made to different nodes and no unique peers process all the requests. 

Reviewer: ```suggestion
- *load balancing*, since requests are made to different nodes and no unique peers process all the requests. 
```

Reviewer: Moved this change to https://github.com/ipfs/docs/pull/364

> ## How do DHTs work?

Reviewer: ```suggestion
## How does the IPFS DHT work?
```

> ## How do DHTs work?

Reviewer: Resolved in https://github.com/ipfs/docs/pull/240/files#r339145095

### Peer IDs

> Each peer have a peerID which is a hash with the same length *n* as the DHT keys.

Reviewer: ```suggestion
Each peer has a `peerID`, which is a hash with the same length *n* as the DHT keys.
```

Reviewer: Moved this change to https://github.com/ipfs/docs/pull/364

### Buckets
A subset of the DHT maintained by a peer is called a 'bucket'. 

> A bucket maps to hashes with the same prefix as the PeerID up to *m* bits. There are 2^m buckets. Each bucket maps for 2^(n-m) hashes.

Reviewer: ```suggestion
A bucket maps to hashes with the same prefix as the `peerID`, up to *m* bits. There are 2^m buckets. Each bucket maps for 2^(n-m) hashes.
```

Reviewer: Moved this change to https://github.com/ipfs/docs/pull/364

> - For example, if m = 2^16 and with an hexadecimal encoding (4 bits per displayed character), the peer with PeerID "ABCDEF12345" maintains mapping for hashes starting with "ABCD". 

Reviewer: ```suggestion
- For example, if *m*=2^16 and we use hexadecimal encoding (four bits per displayed character), the peer with `peerID` 'ABCDEF12345' maintains mapping for hashes starting with 'ABCD. 
```

Reviewer: Moved this change to https://github.com/ipfs/docs/pull/364

> Some hashes falling into this bucket would be *ABCD*38E56, *ABCD*09CBA or *ABCD*17ABB ... 

Reviewer: ```suggestion
Some hashes falling into this bucket would be *ABCD*38E56, *ABCD*09CBA or *ABCD*17ABB, just as examples.
```

Reviewer: Moved this change to https://github.com/ipfs/docs/pull/364

> The size of the buckets are related to the size of the prefix. The longer the prefix, the less hashes each peer has to manage, the more peers are needed.

Reviewer: ```suggestion
The size of a bucket is related to the size of the prefix. The longer the prefix, the fewer hashes each peer has to manage, and the more peers are needed.
```

Reviewer: Moved this change to https://github.com/ipfs/docs/pull/364

Several peers can be in charge of the same bucket if they have the same prefix.

> In most DHTs, including IPFS's Kademlia implementation, the size of the buckets (and the size of the prefix), are dynamic. 

Reviewer: ```suggestion
In most DHTs, including IPFS's [Kademlia implementation](https://github.com/libp2p/specs/blob/8b89dc2521b48bf6edab7c93e8129156a7f5f02c/kad-dht/README.md), the size of the buckets (and the size of the prefix), are dynamic. 
```

Reviewer: Moved this change to https://github.com/ipfs/docs/pull/364

> Buckets size growths and prefix size shortens when many nodes leaves the DHT, and vice versa. **(/!\ or does it depends on the number of records and not the number of nodes? Is it at bucket level or DHT level?)**

Reviewer: ```suggestion

```

Reviewer: Moved this change to https://github.com/ipfs/docs/pull/364

### Peer lists

> Peers also keep connection to other peers to forward requests if the requested hash is not in their own bucket.

Reviewer: ```suggestion
Peers also keep a connection to other peers in order to forward requests if the requested hash is not in their own bucket.
```

Reviewer: Moved this change to https://github.com/ipfs/docs/pull/364

If hashes are of length n, a peer will keep n-1 lists of peers: 

> - the first list contains peers which ID have a different 1st bit.

Reviewer: ```suggestion
- the first list contains peers whose ID have a different first bit.
```

Reviewer: Moved this change to https://github.com/ipfs/docs/pull/364

> - the second list contains peer which have their first bits identical to its own, but a different second bit

Reviewer: ```suggestion
- the second list contains peers whose IDs have first bits identical to its own, but a different second bit
```

Reviewer: Moved this change to https://github.com/ipfs/docs/pull/364

- ...

> - the m-th list contains peer which have their first m-1 bits identical, but a different m-th bit

Reviewer: ```suggestion
- the m-th list contains peers whose IDs have their first m-1 bits identical, but a different m-th bit
```

Reviewer: Moved this change to https://github.com/ipfs/docs/pull/364

- ...

> The higher m, the harder it is to find peers which have the same ID up to m bits. The lists of "closest" peers typically remains empty.

Reviewer: ```suggestion
The higher m is, the harder it is to find peers that have the same ID up to m bits. The lists of "closest" peers typically remains empty.
```

Reviewer: Moved this change to https://github.com/ipfs/docs/pull/364

"Close" here is defined as the XOR distance, so the longer the prefix they share, the closer they are.

> Lists also have a maximum of entries k, otherwise the first lists would contain half the network, then a fourth, etc.

Reviewer: ```suggestion
Lists also have a maximum of entries (k) — otherwise the first lists would contain half the network, then a fourth of the network, and so on.
```

Reviewer: Moved this change to https://github.com/ipfs/docs/pull/364

### Using the DHT

> When a peer receives a lookup request, it will either answer with a value if it falls into its own bucket, or answer with the contacting information (IP + port, peerID, etc) of a closer peer. The requesting peer can then send its request to this closer peer. The process goes on until a peer is able to answer it. 

Reviewer: ```suggestion
When a peer receives a lookup request, it will either answer with a value if it falls into its own bucket, or answer with the contacting information (IP+port, `peerID`, etc.) of a closer peer. The requesting peer can then send its request to this closer peer. The process goes on until a peer is able to answer it. 
```

Reviewer: Moved this change to https://github.com/ipfs/docs/pull/364

A request for a hash of length n will take at maximum log2(n) steps, or even log2m(n). 

> # The DHT of IPFS

Reviewer: ```suggestion
```

Reviewer: Moved this change to https://github.com/ipfs/docs/pull/364

> In IPFS Kademlia's DHT, keys are SHA256 hashes.

Reviewer: ```suggestion
In IPFS's Kademlia DHT, keys are SHA256 hashes.
```

Reviewer: Moved this change to https://github.com/ipfs/docs/pull/364

[PeerIDs](https://docs.libp2p.io/concepts/peer-id/) are those of [libp2p](https://libp2p.io/), the networking library used by IPFS.

> We use a DHT to lookup two types of objects (both represented by a SHA256):

Reviewer: ```suggestion
We use the DHT to look up two types of objects, both represented by SHA256 hashes:
```

Reviewer: Moved this change to https://github.com/ipfs/docs/pull/364

> - [Content IDs](https://docs.ipfs.io/guides/concepts/cid/) of the data added to IPFS. A lookup of this value will give the peerIDs of the peers having this immutable content

Reviewer: ```suggestion
- [Content IDs](https://docs.ipfs.io/guides/concepts/cid/) of the data added to IPFS. A lookup of this value will give the `peerID`s of the peers having this immutable content.
```

Reviewer: Moved this change to https://github.com/ipfs/docs/pull/364

> - [IPNS records](https://docs.ipfs.io/guides/concepts/ipns/). A lookup will give the last Content ID associated with this IPNS address, enabling routing mutable content

Reviewer: ```suggestion
- [IPNS records](https://docs.ipfs.io/guides/concepts/ipns/). A lookup will give the last Content ID associated with this IPNS address, enabling the routing of mutable content.
```

Reviewer: Moved this change to https://github.com/ipfs/docs/pull/364

> Consequently, IPFS's DHT is one of the way of mutable and immutable [Content Routing](https://docs.libp2p.io/concepts/content-routing/). It's currently the only one [implemented](https://libp2p.io/implementations/#peer-routing). 

Reviewer: ```suggestion
Consequently, IPFS's DHT is one of the ways to achieve mutable and immutable [content couting](https://docs.libp2p.io/concepts/content-routing/). It's currently the only one [implemented](https://libp2p.io/implementations/#peer-routing). 
```

Reviewer: Moved this change to https://github.com/ipfs/docs/pull/364

> Per specification, the default bucket size k is 20 meaning each of the 255 lists of peers contain at most 20 peers. 

Author: Is this too much details?

Reviewer: Nope! If it’s too much detail for a reader, it can be skipped over.

Author: Roger that.

> Per specification, the default bucket size k is 20 meaning each of the 255 lists of peers contain at most 20 peers. 

Reviewer: You can learn more in the [libp2p Kademlia DHT specification](https://github.com/libp2p/specs/blob/8b89dc2521b48bf6edab7c93e8129156a7f5f02c/kad-dht/README.md).

Reviewer: Moved this change to https://github.com/ipfs/docs/pull/364

> 

Author: **(what is m and k for IPFS? Does it depends on specs? implementation? )**

Author: Or is it useless in this doc?

Reviewer 2: We should probably just reference the specs. The numbers may vary in the implementation based on performance needs/capabilities, but I believe both go and js are currently using the spec defaults. This may change in the future as the DHT is improved.

## Usage

### Adding an entry

> Adding a blob of data to IPFS is just advertizing that you have it. Since DHT is the only content routing implemented, you can just use

Reviewer: ```suggestion
Adding a blob of data to IPFS is the equivalent fo advertising that you have it. Since DHT is the only content routing implemented, you can just use
```

Reviewer: Moved this change to https://github.com/ipfs/docs/pull/364

` ipfs add myData`

> IPFS will automatically chunk your data and add a mapping on the DHT between the Content ID and your PeerID. Note that there can be other Peer IDs already mapped to that value, so you will be added to the list. Also note that if the provided data is bigger than 124kb, it will be chunked in "blocks" and both blocks and overall data will be mapped. 

Reviewer: ```suggestion
IPFS will automatically chunk your data and add a mapping on the DHT between the Content ID and your `peerID`. Note that there can be other `peerID`s already mapped to that value, so you will be added to the list. Also note that if the provided data is bigger than 124KB, it will be chunked into "blocks", and both those blocks and the overall data will be mapped. 
```

Reviewer: Moved this change to https://github.com/ipfs/docs/pull/364
